### PR TITLE
Support for building DMatrix from Apache Arrow data format

### PR DESCRIPTION
**Pull request by zhangzhang10**
Apache Arrow defines a columnar memory format, which allows zero-copy read and lighting fast data access ([link](https://arrow.apache.org)). This PR adds support for building xgboost DMatrix from the Apache Arrow data format. 

This is a rework of #5667, which is now closed and all discussion should continue here. 

### Features:

- A new adapter class in C++ that accepts input data in the format defined by [the Arrow C data interface ](https://arrow.apache.org/docs/format/CDataInterface.html).
- A new template specialization for the `SimpleDMatrix` class constructor that creates a DMatrix instance from Arrow data.
- No dependence on the Arrow API or the Arrow libraries. It only requires the input data to be in the Arrow format.
- Works with any data source that can export data following the Arrow C data interface specification.
- The xgboost Python interface supports building DMatrix from pyarrow data source, including `pyarrow.Table` and `pyarrow.dataset`. It provides better performance in building DMatrix than using the existing pandas interface. 
- Supports a data iterator interface to allow input data to be provided in a batch-by-batch fashion. This helps to reduce peak memory consumption when reading large data. This also makes it easier to work with other components that expect input data in mini batches, for example, the JVM packages. 
- Tested with Arrow 3.0, 4.0, and 5.0. 

### Performance of building DMatrix (Timing measured on Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz with 56 cores): 

Dataset | File format | Data size | From pandas dataframe | From Arrow format (using pyarrow)|
-- | -- | -- | -- | -- |
HIGGS | csv | 2 million rows, 29 features | 8.43 sec | 0.96 sec |
Mortgage | parquet | 36 million rows, 47 features | 14.31 sec | 2.61 sec |

### Peak memory during DMatrix building (Using a Mortgage dataset with 150 million rows as an example):

Data source | Peak mem |
-- | --  |
pandas.dataframe | 170 GB |
pyarrow.Table | 152 GB |
pyarrow.dataset | 105 GB |

The final DMatrix memory size is 58 GB.  

### How to build:

- Configure the build by passing `-DUSE_ARROW=ON` to the cmake command.

### Usage (python):

A DMatrix can be built from either a `pyarrow.Table` or a `pyarrow.dataset`. For example,
```python
import xgboost as xgb
from pyarrow import csv
import pyarrow.dataset as ds

# Reading CSV input as pyarrow.Table
table = csv.read_csv('/path/to/csv/input')
dmat = xgb.DMatrix(table)

# Reading parquet input as pyarrow.dataset
data = ds.dataset('/path/to/parquet/intput', format='parquet')
dmat = xgb.DMatrix(data)
```

### CODE REVIEW

File: python-package/xgboost/core.py [hunk 1]
```diff
@@ -606,13 +636,44 @@ def __init__(
...
 
-        from .data import dispatch_data_backend, _is_iter
+        from .data import dispatch_data_backend, _is_iter, _is_arrow
 
         if _is_iter(data):
             self._init_from_iter(data, enable_categorical)
             assert self.handle is not None
             return
 
+        if _is_arrow(data):
```
[the remark this answers is not in the document]
Comment (zhangzhang10) [reply python-package/xgboost/core.py:620]: > @zhangzhang10 Do you think we can work on the smaller subset of this PR first like getting the predictor `X`?

By a smaller subset, do you mean as the first step we only support the predictor `X` (i.e. the feature columns) as arrow format, and treat metainfo columns as array-like objects? If so, then yes I agree. I'm working on this. Thanks.

File: src/data/simple_dmatrix.cc [hunk 2]
```diff
@@ -236,5 +236,142 @@ template SimpleDMatrix::SimpleDMatrix(
     IteratorAdapter<DataIterHandle, XGBCallbackDataIterNext, XGBoostBatchCSR>
         *adapter,
     float missing, int nthread);
+
+#if defined(XGBOOST_BUILD_ARROW_SUPPORT)
+template<>
+SimpleDMatrix::SimpleDMatrix(RecordBatchesIterAdapter* adapter,
```
[the remark this answers is not in the document]
Comment (zhangzhang10) [reply src/data/simple_dmatrix.cc:242]: Per discussion, I've revised this PR such that we stick with the existing `CreateDMatrixFromXXX` + `SetInfo` route for DMatrix initialization. When using the Python API, meta info such as labels and weights are set separately using `SetInfo`. This is consistent with the sklearn custom.

File: python-package/xgboost/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhangzhang10) [reply python-package/xgboost/core.py]: @trivialfis Done. This PR currently only builds predictor DMatrix `X` from arrow data.

**Comment by zhangzhang10** [comment]
> I think for step 1 we need to decouple the predictor `X` from the rest of the meta info. Feel free to share your thoughts.

Yes. This step is done.

File: src/data/adapter.h [hunk 3]
```diff
@@ -760,6 +763,523 @@ class IteratorAdapter : public dmlc::DataIter<FileAdapterBatch> {
...
+  size_t length() const { return length_; }
+
+ protected:
+  size_t col_idx_;
+  size_t length_;
+  size_t null_count_;
+  const uint8_t* bitmap_;
+};
+
+// only columns of primitive types are supported
```
[the remark this answers is not in the document]
Comment (zhangzhang10) [reply src/data/adapter.h]: I've added some comment for clarification.

**Comment by trivialfis** [comment]
Thanks!  I will look into the details next week.  Really appreciate your excellent work!

**Comment by trivialfis** [comment]
I have restarted the CI.

**Comment by FelixYBW** [comment]
> 

@trivialfis any work we need to do?

**Comment by trivialfis** [comment]
I haven't looked into your new code yet, but from the CI:

```
[2021-11-15T22:21:09.849Z] tests/python/test_updaters.py::TestTreeMethod::test_hist_degenerate_case PASSED

[2021-11-15T22:21:10.105Z] tests/python/test_with_arrow.py::TestArrowTable::test_arrow_dataset_from_csv FAILED

[2021-11-16T01:39:20.757Z] tests/python/test_with_arrow.py::TestArrowTable::test_arrow_survival Sending interrupt signal to process

[2021-11-16T01:39:31.604Z] script returned exit code 143
```

**Comment by xwu99** [comment]
Hi @trivialfis
Sorry for late reply. @zhangzhang10 has handed the work over to me. I will create a seperate PR to work on this.

**Comment by pseudotensor** [comment]
How's progress @xwu99 ?

**Comment by hcho3** [comment]
@pseudotensor Work is being continued in #7512
